s3-test/src/main/java/com/hazelcast/jet/tests/s3/S3WordCountTest.java

```diff
@@ -104,33 +103,30 @@ protected void test() {
                 jobConfig.setName("s3-test-" + jobNumber);
                 jet.newJob(pipeline(), jobConfig).join();
                 verify(jobNumber);
-                jobNumber++;
             } catch (Throwable e) {
-                if (isSocketTimeoutException(e)) {
+                if (isCausedBySocketTimeoutException(e)) {
                     logger.warning("Socket timeout ", e);
                     socketTimeoutNumber++;
                 } else {
                     throw ExceptionUtil.rethrow(e);
                 }
             }
+            jobNumber++;
         }
         long thresholdForSocketTimeout = TimeUnit.MILLISECONDS.toHours(durationInMillis) + 1;
         assertTrue("Socket timeout number is too big", thresholdForSocketTimeout > socketTimeoutNumber);
         logger.info(String.format("Total number of jobs finished: %d, socketTimeout: %d", jobNumber, socketTimeoutNumber));
     }
 
-    private boolean isSocketTimeoutException(Throwable e) {
+    private boolean isCausedBySocketTimeoutException(Throwable e) {
+        if (e == null) {
+            return false;
+        }
         if (e instanceof SocketTimeoutException) {
             return true;
+        } else {
+            return isCausedBySocketTimeoutException(e.getCause());
         }
-        if (e instanceof UndefinedErrorCodeException) {
```

Comment:
when the exception happens on server-side it is returned to the client in `UndefinedErrorCodeException` afaik. So the cause is never an instance of `SocketTimeoutException`, this line was checking for that. What I see from last stack-trace we should also check for `SocketException`